[New Block] Add post time to read block

t-hamano opened the pull request:
Close #42542

_Note_: The latest status is summarized in [this comment](https://github.com/WordPress/gutenberg/pull/43403#issuecomment-1277607866).

## What?
This PR adds a new "**Post Time To Read**" block.

For now, I'm keeping implementation to a minimum. If it makes sense to incorporate this block, I would like to work on additional implementation.

## Why?
With #41611, "Time to read" has been added to the Detail panel.
By using the same information in the block, the site owner can tell visitors in advance how long it will take to finish reading.

## Implementation Details

### Calculation of time to read
The time required indicated by the block must match the "Time to read" in the Details panel. This PR uses the same `@wordpress/wordcount` package to calculate.

The `count` method of this package is very powerful and supports word-by-word or character-by-character calculations. The calculation criteria can be controlled by the translation function, so the time required can be calculated appropriately according to the site language:

https://github.com/WordPress/gutenberg/blob/01611823eabb937a36de6efcf9902677e73e0639/packages/block-library/src/post-time-to-read/edit.js#L39-L44

~~This block stores only the calculated time (`minutesToRead`) in the attribute and outputs the appropriate string through dynamic rendering.~~

 _@update_: Time is no longer stored as an attribute since time is now computed by dynamic rendering.

### ~~Why dynamic rendering?~~

~~When the block was saved as static markup by using `__()`, the block could be broken if the site language is changed, as mentioned in #43050.~~
~~Therefore, We need to use the translation function on the server side.~~

### ~~About demerit~~

~~The biggest disadvantage of this implementation is that time is stored as attributes. The time displayed on the front end is updated only when the post is saved.~~

~~This means that it will not work properly in templates such as `single.html` or in query loops. This issue is similar to what was mentioned in #29739 when the TOC block was rebuilt with static markup.~~

 _@update_: Time is now fully computed by dynamic rendering. Therefore, it works not only in the post editor, but also in the site editor templates.

## Testing Instructions

First, try using this block on gutenberg.run.

http://gutenberg.run/43403

## Next steps and suggestions

If it makes sense to bundle this block, I would like to proceed with this PR according to the following steps and suggestions.

### Add E2E tests
We will need to check that the correct attributes (`minutesToRead`) have been saved after inserting content of any length.

### Add block supports
As discussed in #43241, we need to add all the necessary block support to address Design Tooling Consistency.

### Add filter hook

The time string should be displayed appropriately depending on the language of the site, but it might make sense to provide hooks such as the following:

```php
$minutes_to_read_string = apply_filters( 'block_core_post_time_to_read_render_string', $minutes_to_read_string );
```

### ✅ Create a block icon (Done)

I could not find a suitable icon from the `@wordpress/icons` package, so I created a temporary icon.
Need help from designers to improve this icon or create a completely new icon 🙏

_@update_: As [this comment](https://github.com/WordPress/gutenberg/pull/43403#issuecomment-1265252253) states, @jasmussen provided the icons 🎉

### ✅ Server-side time calculation (Done)

As mentioned earlier, this block will only work on a single post.

Many consumers will probably want to place the block in a template such as `single.html`. To do so, I think we would need to implement the exact same logic on the server side as `@wordpress/count` does.

## Screenshots or screencast

![editor_1](https://user-images.githubusercontent.com/54422211/190885160-9962e7e1-8381-4c85-915b-96fcdcc2eb1d.png)
![editor_2](https://user-images.githubusercontent.com/54422211/190885162-867143c5-a0d6-46cb-bc6c-d2c3d6812f9e.png)
![front](https://user-images.githubusercontent.com/54422211/190885163-1c4c4298-0a48-49ea-a945-78f6ebec16bb.png)

t-hamano committed:
Add new "Post Time To Read" block

t-hamano committed:
Add fixture

t-hamano committed:
Update npm dependencies

t-hamano committed:
Disallow multiple insert

t-hamano committed:
Fix php lint error

t-hamano committed:
Fix npm dependencies path

t-hamano committed:
Remove unused context

bph commented:
That's a fun block! Thank you. 

I added it to the single post via the post editor. 

![Screen Shot 2022-09-21 at 16 25 21](https://user-images.githubusercontent.com/39980/191603840-d38514d3-15dd-43db-bfd7-e7a551567ef1.png)

For my test, I also wanted to use it on my home page with the post grid, now I got different values: 
The site editor show "no content' for the block. 
![Screen Shot 2022-09-21 at 16 26 01](https://user-images.githubusercontent.com/39980/191604108-b97d7623-019d-4af6-9286-23f4aec066b0.png)

On the front end is shows "less than a minute" which is obviously different from the 9 min on the single post. 

![Screen Shot 2022-09-21 at 16 26 17](https://user-images.githubusercontent.com/39980/191604203-2f24b2a4-1e1c-4041-a5b0-3c85f12f9ed7.png)

There is also a missing 'in' there: "You can read this post **in** less than a minute"

Maybe there is a way to get the value stored in the post_meta table so it can be used like any other post_meta information, like post_date, or post_author.

t-hamano committed:
Add a missing word in translate string

t-hamano commented:
Thank you for the review, @bph !

> There is also a missing 'in' there: "You can read this post in less than a minute"

Fixed 👍

> Maybe there is a way to get the value stored in the post_meta table so it can be used like any other post_meta information, like post_date, or post_author.

I would like to hear from the developers if it is possible to add new meta information when saving the post and if this would have any significant impact. Also, when someone moves from the classic editor to Gutenberg, there will be no meta information initially.

I will ask for feedback once in the editor chat.

talldan commented on packages/block-library/src/post-time-to-read/edit.js:
```diff
@@ -0,0 +1,96 @@
+	const content = useSelect(
+		( select ) => select( editorStore ).getEditedPostAttribute( 'content' ),
+		[]
+	);
```
It'd be good to look at whether it's possible to switch to `useEntityProp`, as that would make it work in the site editor. You can look at how a block like the Post Title block works for inspiration, but I think it would be something like this:
```js
export default function PostTimeToReadEdit( {
	attributes,
	setAttributes,
	context: { postType, postId },
} ) {

	const [ content = '' ] = useEntityProp(
		'postType',
		postType,
		'content',
                postId
	);
```

t-hamano replied:
Fixed by 4118728

talldan commented on packages/block-library/src/post-time-to-read/edit.js:
```diff
@@ -0,0 +1,96 @@
+		__unstableMarkNextChangeAsNotPersistent();
+		setAttributes( {
+			minutesToRead: content ? newMinutesToRead : undefined,
+		} );
```
If there is a way to avoid storing this as an attribute, I think that would be a positive thing to do. The block could be entirely deterministic based on the entity.

My head is telling me there would be two main tasks:
- Make the block entirely dynamic
- Port wordCount to PHP (or use existing code if there's already an equivalent word count PHP funciton).

t-hamano replied:
I have tried not to save the `minutesToRead` by 92190e0.

As you say, if we could implement a complete PHP function, we would not need attributes, as they can be calculated from the content.

talldan commented on packages/block-library/src/post-time-to-read/edit.js:
```diff
@@ -0,0 +1,96 @@
+				? sprintf(
+						/* translators: %d is the number of minutes the post will take to read. */
+						_n(
+							'You can read this post in %d minute.',
+							'You can read this post in %d minutes.',
```
I think the text could be configurable, as it's unlikely to suit everyone. I'm not sure if there are any other blocks that have a way to configure text where there's a variable involved though. That seems like a challenge!

t-hamano replied:
I initially considered introducing a prefix and suffix like `core/post-terms`.

![time-to-read](https://user-images.githubusercontent.com/54422211/193535742-9a78b7ec-a8a7-4b01-bf28-5cde27074456.png)

However, a common string would not cover all cases of singular and plural, or all cases where the time required is less than a minute.

I also considered the following sidebar setting, but this is complicated 🤔

![settings](https://user-images.githubusercontent.com/54422211/193535958-efc4f907-3b88-49ff-977c-a4c19964dd76.png)

talldan replied:
I thought the easiest way to start might be just a prefix, and making the unit of time an implicit part of the text that can't be changed. The only confusing thing would be how internationalization works. User entered text can't be localized at the moment, but the 'minute' or 'minutes' would be. 🤔 

Lets see if we can get some assistance from a designer on this particular matter (cc @WordPress/gutenberg-design).

jameskoster replied:
Maybe the block should just print the time, then if you want a prefix/suffix you can place it in a Row alongside Paragraphs?

t-hamano replied:
> Maybe the block should just print the time, then if you want a prefix/suffix you can place it in a Row alongside Paragraphs?

Yes, I think such an approach is possible.

However, given that [the core/post-terms block has a prefix/suffix](https://github.com/WordPress/gutenberg/blob/dbd48b6f3a6085806f2bf0ec4c28b19f5e02f563/packages/block-library/src/post-terms/block.json#L20-L27), I think the current approach makes sense. Also, if this block displayed only time, adding a prefix or suffix would create multiple paragraphs. I think it should consist of a single paragraph, including prefixes and suffixes.

jameskoster replied:
Good point.

> The only confusing thing would be how internationalization works. User entered text can't be localized at the moment

If the prefix/suffix had a placeholder, would that be localized? If so I believe that would solve any issues?

t-hamano replied:
> If the prefix/suffix had a placeholder, would that be localized? If so I believe that would solve any issues?

Yes, the prefix/suffix placeholder will be localized as you can see [here](https://github.com/WordPress/gutenberg/blob/fb5f3c73f3e6779ae1aa1a8e511cb8ea46fef7a3/packages/block-library/src/post-time-to-read/edit.js#L128-L129).

However, the prefix/suffix with text entered are stored as attributes and aren't localized.
Therefore, for example, if you enter prefix / suffix in English and switch to a language other than English, only the time will be localized and the correct text will not be generated.
In early versions, there was [an implementation to make all paragraphs subject to localization](https://github.com/WordPress/gutenberg/blob/689894757aedbcae4275f3037fc44ebdef9c466d/packages/block-library/src/post-time-to-read/index.php#L24-L34), without prefix/suffix. However, this would be inconvenient because the user wouldn't be able to change the text. (This may be solved by introducing filter hooks.)

talldan commented on packages/block-library/package.json:
```diff
@@ -44,6 +44,7 @@
 		"@wordpress/date": "file:../date",
 		"@wordpress/deprecated": "file:../deprecated",
 		"@wordpress/dom": "file:../dom",
+		"@wordpress/editor": "file:../editor",
```
I don't think `@wordpress/editor` should be introduced as a dependency to `@wordpress/block-library`. I know it is used in a couple of places (calender, table of contents, template part), but I think introducing this as an explicit dependency was always avoided.

Reasoning - the dependency chain is conceptually the other way around (editor uses blocks), so it's a bit risky in terms of the chance of introducing a circular dependency.

Other blocks use the 'core/editor' string to avoid the package dependency and treat it as an optionally available store. There's also usually a comment about the problem. Example here:
https://github.com/WordPress/gutenberg/blob/a6e9a24798ac4e8ca954ebdf762a614a2c4df7ff/packages/block-library/src/calendar/edit.js#L58-L70

Switching to `useEntityProp` would be good to explore though.

t-hamano replied:
By rewriting `getEditedPostAttribute` to `useEntityProp`, the import of this package is no longer necessary 👍

talldan reviewed (commented):
Nice work here ❤️

t-hamano commented:
Thank you for the review, @talldan !
I would like to correct the points you have commented on.

> My head is telling me there would be two main tasks:
> 
> Make the block entirely dynamic
> Port wordCount to PHP (or use existing code if there's already an equivalent word count PHP funciton).

Yes, I think this is a major point. I think it would be ideal to implement this block as a dynamic block and work in a template (or in a query loop block). However, as far as I know, there is no function like `@wordpress/count` in the WordPress core.

If it were to be implemented in PHP, I think the following step would be a good idea, what do you think?

- Implement the complete logic within the Time To Read callback function
- Backport this function to the WordPress core (could be a function like `the_time_to_read( $str )` or `get_the_time_to_read( $str )`)
- Replace the logic in render_callback with this function backported into the WordPress core

talldan commented:
> If it were to be implemented in PHP, I think the following step would be a good idea, what do you think?

Yep, something like that sounds great. You can implement the function in the `lib/compat/wordpress-6.2` folder (no folder exists at the moment, but feel free to create one). This is the folder that will contain all the code that needs to be backported for WordPress 6.2. Functions implemented in this file typically have the `gutenberg_` prefix, which is then removed for the core version of the function when backporting. The new file can be required in the `lib/load.php` file. It is good to get an idea of the where the function will be added to in the core codebase and leave that in a comment. Possibly `functions.php` as that has some related code like the `human_readable_duration` function, but I'm no expert about this kind of thing.

An alternative is to implement everything in the block's index.php file, but one thing to note is that the backporting process is different. The php files in the block library package are automatically copied to the core codebase, there's no manual backport. Because of that you won't be able to use the `gutenberg` prefix as that would be copied to core, so it can be a bit of a problematic way to implement additional PHP functions.

t-hamano committed:
Don't use @wordpress/editor package

t-hamano committed:
Don't store  attribute

t-hamano committed:
Fix npm dependencies

t-hamano commented:
Thank you for the review, @talldan !

> Make the block entirely dynamic
> Port wordCount to PHP (or use existing code if there's already an equivalent word count PHP funciton).

Yes, this is probably the biggest problem here.
In the WordPress core, [wp-includes/formatting.php](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/formatting.php) seems to be relevant, but there doesn't seem to be an alternative function for `wordCount`, so I think we'll have to create a completely new function.
This is complicated, but I'm willing to give it a shot 👍

jasmussen commented:
This is a fun little utility block! I took it for a spin:

![time to read](https://user-images.githubusercontent.com/1204802/193554713-9f2154d2-0b27-4616-a581-740dea696113.gif)

It works well. The crash happens most easily if you select the block and press Enter, though it happened also seemingly a bit at random. Got this in the console:

```
react-dom.min.js?ver=6.1-beta2-54366:141 TypeError: text.replace is not a function
    at stripTags (stripTags.js:10:14)
    at index.js:76:30
    at Array.reduce (<anonymous>)
    at countWords (index.js:76:4)
    at count (index.js:124:11)
    at edit.js:48:14
    at Tj (react-dom.min.js?ver=6.1-beta2-54366:179:399)
    at unstable_runWithPriority (react.min.js?ver=6.1-beta2-54366:24:26)
    at Za (react-dom.min.js?ver=6.1-beta2-54366:73:8)
    at Ra (react-dom.min.js?ver=6.1-beta2-54366:178:368)
```

The icon looks pretty good. I suspect a clock is useful enough that more blocks, or even parts of the UI could have this. But since it seems appropriate for the block at hand, let's use it! You can probably add it to the icon library, though. The metrics were pretty good, but here's a new version that matches similar circle icons:
<img width="420" alt="Screenshot 2022-10-03 at 12 28 07" src="https://user-images.githubusercontent.com/1204802/193556495-01a9be33-7ad8-40a4-8663-0f7a494dfbbc.png">

```
<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" style="enable-background:new 0 0 24 24" xml:space="preserve"><path d="M12 3c-5 0-9 4-9 9s4 9 9 9 9-4 9-9-4-9-9-9zm0 16.5c-4.1 0-7.5-3.4-7.5-7.5S7.9 4.5 12 4.5s7.5 3.4 7.5 7.5-3.4 7.5-7.5 7.5zM12 7l-1 5c0 .3.2.6.4.8l4.2 2.8-2.7-4.1L12 7z" style="fill:#1e1e1e"/></svg>
```

Let me know if that works for you. 

Question regarding [this conversation](https://github.com/WordPress/gutenberg/pull/43403#discussion_r985522997): instead of including the full text of "This post can be read in n minutes", which runs into translation and theoretically even post type challenges, could we literally output _only_ the time to read? I.e. 

* 3 minutes
* 1 minute

Perhaps even with a toggle to choose between the full `minutes` and the shorter `m` (3m, etc), the idea being you could type out any text you like in a paragraph, reducing the burden on this block. I realize that opens a conversation on blocks vs. inline blocks, and how it remains tricky to create a horizontal row of post meta (a la #26899). 

On the other hand, that appears to be partially handled by prefix/suffixes also (#40559), so perhaps that route is the way to go after all? Something to consider.

paaljoachim commented:
I too received a "This block has encountered an error and cannot be previewed." error.
With a similar console error code. I added the block with the forward slash. / The error was seen right after adding the block.
Good to see that it can only be added once to any post/page (?).

t-hamano committed:
Fix: The block is broken in the post editor

t-hamano committed:
Update the block icon

t-hamano commented:
Thanks for the feedback, @jasmussen, @paaljoachim!
I have fixed the problem of broken blocks in the post editor.

@talldan 
I have found the logic to retrieve the current (unsaved) content without using the `core/editor` package.
https://github.com/WordPress/gutenberg/blob/af5802af9655d3af59b3293f9053f27b5b8f3dd4/packages/edit-site/src/components/code-editor/index.js#L36-L47

Both the post editor and the site editor will retrieve the content correctly.

t-hamano commented:
@jasmussen 
I have replaced the icon. I think it works very well!

![icon_preview](https://user-images.githubusercontent.com/54422211/193577934-cc53bdac-8b74-4688-b29d-40f33e4fd98f.png)

t-hamano committed:
Add prefix and suffix

t-hamano commented:
I have added the prefix and suffix.

![preview](https://user-images.githubusercontent.com/54422211/193583207-3354a231-96df-40dd-a6e4-1b6ebca4a96f.png)

At least in English, I believe that any variation can produce the correct sentence. I think there might be no problem if this text is properly translated:

https://github.com/WordPress/gutenberg/blob/5221b9d58297769e4dc98df377d5515a7dd5409c/packages/block-library/src/post-time-to-read/edit.js#L91-L96

<details>
<summary>Without prefix and suffix</summary>

![without_prefix_suffix](https://user-images.githubusercontent.com/54422211/193585317-1314979f-b895-475f-81de-eb9a24ce6883.png)

</details>

<details>
<summary>With prefix and suffix</summary>

![with_prefix_suffix](https://user-images.githubusercontent.com/54422211/193585329-5b4bf874-0ea8-4e81-8911-b85433dbe63b.png)

</details>

t-hamano committed:
Server-side rendering implemented using dummy function

t-hamano committed:
Update fixtures

t-hamano committed:
Fix phpcs lint error

t-hamano committed:
Implement word_count function

t-hamano committed:
Remove some new lines on php

t-hamano commented:
@talldan 
I was able to implement wordcount in PHP 😊

Perhaps the original logic is here:
https://github.com/WordPress/wordpress-develop/blob/2648a5f984b8abf06872151898e3a61d3458a628/src/js/_enqueues/wp/utils/word-count.js

This is probably what `@wordpress/wordcount` was implemented based on (Actually, the logic is exactly the same.). 
I think it would be better to get a review from the core team on this new function implemented in PHP, what do you think?

t-hamano committed:
Fix regexp escape

t-hamano commented:
@bph 
This block now also works in the site editor 🚀 I would be happy to test it when you have time.

paaljoachim commented:
Retesting using
http://gutenberg.run/43403

I see this:

![Screenshot 2022-10-04 at 22 16 51](https://user-images.githubusercontent.com/5323259/193918089-0caae92d-280f-48c8-b519-01a4ef5ce27a.jpg)

Prefix "less than a minute" Suffix.

**Should Suffix be with a small s?**

I added a prefix of "It will take" and the Suffix "to read this text."
![Screenshot 2022-10-04 at 22 19 18](https://user-images.githubusercontent.com/5323259/193918322-baeeada8-17e2-4a77-afef-e6a57c7b3181.jpg)

Result on the frontend.
![Screenshot 2022-10-04 at 22 21 18](https://user-images.githubusercontent.com/5323259/193918736-84767820-e84b-4501-8c91-cd27b9c1078d.jpg)

Backend text says:
"It will take less than a minute to read this text."

Frontend text says:
"It will take 0 minutes to read this text."

**The same backend text should also be seen on the frontend.**

t-hamano committed:
convert float zero to int zero

t-hamano commented:
Thanks for the retest, @paaljoachim !

> The same backend text should also be seen on the frontend.

Fixed 👍

> Should Suffix be with a small s?

I referred to `core/post-terms` (Tags , Categories) for the prefix and suffix implementation, but it seems to be a uppercase beginning Suffix.

![post-term](https://user-images.githubusercontent.com/54422211/193949026-4b3346a0-0579-4b13-b022-331686e58df5.png)

t-hamano committed:
Fix PHP lint error

talldan commented on packages/block-library/src/post-time-to-read/edit.js:
```diff
@@ -0,0 +1,155 @@
+						_n( '%d minute', '%d minutes', minutesToRead ),
+						minutesToRead
+				  )
+				: __( 'less than a minute' );
```
It might be an idea to capitalize this to sentence case ('Less than a minute') whenever the prefix is empty.

t-hamano replied:
Fixed by b14aca6

paaljoachim commented:
Good work Aki!

Another test with Gutenbergrun.
In Brave
On a Mac.

Clicking into the Prefix less than a minute Suffix can be tricky.
I have to click after minute to be able to have the cursor blink inside it. (Making it editable).

![Selectable clickable area](https://user-images.githubusercontent.com/5323259/193994377-9c486507-51c9-4c29-9d9e-47f19d3ef994.jpg) 

Clicking inside the pink area I am able to see the cursor blink that gives me the signal that I can edit something.

https://user-images.githubusercontent.com/5323259/193994821-b93ba8ad-a572-43c0-afd7-61345a168514.mp4
I used the keyboard arrow keys to move the cursor back to the Prefix area. This was difficult. I could not see the cursor move backward and it suddenly jumped up to the title. (Can really not be seen in the video.)

Clicking into Prefix or "less than a minute" is not possible. Clicking into these areas that cursor should be seen just before Prefix (at the beginning of the block). Currently it is not clickable.

Fixed: Backend text is now seen on the frontend.

talldan commented on packages/block-library/src/post-time-to-read/edit.js:
```diff
@@ -0,0 +1,155 @@
+
+	const minutesToRead = Math.round(
+		wordCount( content, wordCountType ) / AVERAGE_READING_RATE
+	);
```
A `useMemo` hook might be good around this code, as I think calling `contentStructure()` and `wordCount()` could be quite expensive.

At the moment it would recalculate on every render, even when the content hasn't changed. For example changing the alignment could cause a recalculation.

t-hamano replied:
Fixed by 7b851e2
Since alignment changes the markup, this process would be performed. However, performance has improved because changes in block selection are no longer recalculated.

talldan commented on lib/compat/wordpress-6.2/formatting.php:
```diff
@@ -0,0 +1,113 @@
+ * }
+ * @return int The word or character count.
+ */
+function gutenberg_word_count( $text, $type, $settings = array() ) {
```
I'm not PHP expert, but this looks very thorough! Good job.

A good next step would be to add unit tests. It might be easiest to replicate the JavaScript unit tests as a starting point if this works in the same way.

t-hamano replied:
Perhaps we need a PHP unit test that is equivalent to this test:
https://github.com/WordPress/wordpress-develop/blob/trunk/tests/qunit/wp-admin/js/word-count.js

But where should I place the test file to test the function stored in `lib/compat`?

talldan commented on packages/block-library/src/post-time-to-read/block.json:
```diff
@@ -0,0 +1,27 @@
+			"default": ""
+		}
+	},
+	"supports": {
+		"html": false,
+		"multiple": false
+	}
```
As a follow-up PR once this is merged, it'd be good to consider what design tools this needs. There are some tracking issues for design tool consistency - https://github.com/WordPress/gutenberg/issues?q=is%3Aopen+is%3Aissue+label%3A%22%5BFeature%5D+Design+Tools%22+label%3A%22%5BType%5D+Tracking+Issue%22.

talldan reviewed (commented):
It's looking really good so far. Great work!

It's nice to see it working so well in post templates in both the post and site editor, as I think that will be a very common way to add the block.

t-hamano committed:
Cursor isn't displayed in prefix and suffix input area

t-hamano commented:
@paaljoachim 

> Clicking into the Prefix less than a minute Suffix can be tricky.
> I have to click after minute to be able to have the cursor blink inside it. (Making it editable).

I found that this is caused by the `contenteditable` attribute being assigned to the inline `span` element. I applied `inline-block` with e4fcde2 and it seems to work correctly.

https://user-images.githubusercontent.com/54422211/194008349-eea84ed8-42c5-412d-80fc-b6aa1fc87ce7.mp4

This problem also occurs in the `core/post-terms` block with the same prefix and suffix input areas and I think it may need to be corrected as well.

t-hamano committed:
Calculate time only when content is changed

t-hamano committed:
Capitalize when there is no prefix

t-hamano committed:
Apply inline block style to editor only

t-hamano committed:
Add Unit Test

t-hamano commented:
Update: It may not be the proper directory, but I have added PHP unit tests [here](https://github.com/WordPress/gutenberg/blob/add/block-post-time-to-read/phpunit/formatting-test.php). This replaces [WordPress core JavaScript tests](https://github.com/WordPress/wordpress-develop/blob/trunk/tests/qunit/wp-admin/js/word-count.js) with PHP.

However, PHP and JavaScript have different Unicode character counting behavior and some tests do not pass, so they are commented out. I'm continuing to work on it, but for example, [U2026](https://www.fileformat.info/info/unicode/char/2026/index.htm) would cause PHP to count it as multiple characters, as shown below.

```JavaScript
// JavaScript
const matches = '…'.match( /\S/g );
console.log( matches.length );
// => 1
```

```PHP
// PHP
preg_match_all( '/\S/', '…', $matches );
echo count( $matches[0] );
// => 3
```

t-hamano committed:
Fix PHP Lint

mtias commented on packages/block-library/src/post-time-to-read/block.json:
```diff
@@ -0,0 +1,27 @@
+	"$schema": "https://schemas.wp.org/trunk/block.json",
+	"apiVersion": 2,
+	"name": "core/post-time-to-read",
+	"title": "Post Time To Read",
```
I think this should be just "Time to Read"

t-hamano replied:
Fixed by cca0f25

mtias commented on packages/block-library/src/post-time-to-read/block.json:
```diff
@@ -0,0 +1,27 @@
+	"name": "core/post-time-to-read",
+	"title": "Post Time To Read",
+	"category": "theme",
+	"description": "Time to read the post.",
```
The description could be expanded a bit

t-hamano replied:
Fixed by https://github.com/WordPress/gutenberg/commit/cca0f253676fd2bd5987ef4d5f6537413266fae4

mtias commented:
I don't think we should support prefix / suffix here. It's an awkward pattern. The right solution for inline dynamic values is along the lines of https://github.com/WordPress/gutenberg/discussions/39831

I think this block should be just what the block outputs initially.

anton-vlasenko commented on phpunit/formatting-test.php:
```diff
@@ -0,0 +1,105 @@
+/**
+ * Tests_Formatting_Wordcount class
+ *
+ * @package WordPress
```
```suggestion
 * @package WordPress
 *
 * @covers ::gutenberg_word_count
```

t-hamano replied:
Fixed by 2f12db3

anton-vlasenko commented on phpunit/formatting-test.php:
```diff
@@ -0,0 +1,105 @@
+				'characters_including_spaces' => 13,
+			),
+
+			/*
```
If this code is not being used, it's better to remove it.
https://github.com/jupeter/clean-code-php#remove-dead-code

t-hamano replied:
I had temporarily disabled this test because I just could not get it to pass.
Resolved by 34dd05f 👍

anton-vlasenko commented on phpunit/formatting-test.php:
```diff
@@ -0,0 +1,105 @@
+			),
+
+			/*
+			// phpcs:ignore Squiz.Commenting.BlockComment.NoCapital
```
Same here.
https://github.com/jupeter/clean-code-php#remove-dead-code

t-hamano replied:
Resolved by 34dd05f

anton-vlasenko commented on phpunit/formatting-test.php:
```diff
@@ -0,0 +1,105 @@
+			),
+		);
+
+		foreach ( $expected_settings as $expected_setting ) {
```
I think using a data provider instead of `foreach` loop is better.
Why?
1. Data providers support named data sets, so it's easier to understand what went wrong in case of an error.
2. Cleaner code. Test data doesn't need to stored within the test method.

t-hamano replied:
Thank you for the detailed explanation.
I have written very little PHPUnit, but I have updated with 885e921.
If you don't mind, I would be happy to review this change to see if it is appropriate.

t-hamano commented:
Thank you for the review, @mtias !

> I don't think we should support prefix / suffix here. It's an awkward pattern. The right solution for inline dynamic values is along the lines of https://github.com/WordPress/gutenberg/discussions/39831
> 
> I think this block should be just what the block outputs initially.

The concepts discussed in #39831 were very difficult for me 😅
My understanding is that this block shouldn't accept user text input because approaches such as dynamic attributes and tokens have not yet been established. Am I correct?

If so, should all text be output dynamically, as in the initial implementation?

https://github.com/WordPress/gutenberg/blob/689894757aedbcae4275f3037fc44ebdef9c466d/packages/block-library/src/post-time-to-read/index.php#L24-L34

mtias commented:
Indeed, we should keep it simple!

t-hamano committed:
Revert prefix/suffix

t-hamano committed:
Change block title and description

t-hamano commented:
Update: Localized entire text without prefix/suffix.

t-hamano committed:
Fix some unit tests

t-hamano committed:
Show message when there is no content

anton-vlasenko commented on lib/compat/wordpress-6.2/formatting.php:
```diff
@@ -0,0 +1,113 @@
+ * }
+ * @return int The word or character count.
+ */
+function gutenberg_word_count( $text, $type, $settings = array() ) {
```
If the function is intended to be merged into Core, then it should start with the `wp_` prefix.
Otherwise,
```
@since 6.2.0
```
should be removed.
https://github.com/WordPress/gutenberg/blob/trunk/lib/README.md

t-hamano replied:
Fixed by ee2f392

anton-vlasenko commented on packages/block-library/src/post-time-to-read/index.php:
```diff
@@ -0,0 +1,75 @@
+ * @param  WP_Block $block      Block instance.
+ * @return string Returns the rendered post author name block.
+ */
+function render_block_core_post_time_to_read( $attributes, $content, $block ) {
```
It would be nice to write a PHPUnit test for this function.
Thanks!

t-hamano replied:
Thanks for the advice. I will give it a try 👍

t-hamano replied:
Added by cd780a2

anton-vlasenko reviewed (dismissed):
Thanks for the PR!
I've reviewed the code and added my notes.

t-hamano committed:
Pass all PHP unit tests

t-hamano committed:
Add @covers to Unit Test class

t-hamano committed:
Changed prefix from gutenberg_ to wp_

t-hamano committed:
Use dataProvider

t-hamano committed:
Add render_callback function unit tests

t-hamano committed:
Add function existence check

t-hamano commented:
I would like to summarize  here our discussions and responses to date.

If you have any concerns as we move forward with this PR, please feel free to comment. I appreciate your support and advice 🙂

## implementation

![capture](https://user-images.githubusercontent.com/54422211/195606909-2879d6bb-1e10-4350-b7a0-b4547c14299c.png)

### Basic

This block calculates the time it takes to finish reading a post and displays the sentences in a single paragraph. Times are calculated dynamically and work both in the post and in the template. Two approaches were considered with respect to the display of text.

- Only the time (less than 1 minute, 2 minutes, etc.) should be displayed dynamically, and the prefix/suffix text can be changed by RichText.
- The entire text is generated dynamically and does not accept user input.

Based on [this comments](https://github.com/WordPress/gutenberg/pull/43403#issuecomment-1275871011), we adopted the latter.

The text is localized and has three variations:

- _You can read this post in less than 1 minute._
- _You can read this post in %d minute._ (singular form)
- _You can read this post in %d minutes._ (plural form)

A future follow-up would be to add block support, as discussed in #43241.

### New function

On the editor, `@wordpress/wordcount` is used to calculate the time to read. On the front-end, the time must be calculated by PHP in the render callback. The logic of this package has been rewritten in PHP. Additionally, `@wordpress/wordcount` is copied from the core [WordCounter](https://github.com/WordPress/wordpress-develop/blob/trunk/src/js/_enqueues/wp/utils/word-count.js) logic. This PHP function will be a new core function called `wp_word_count`. The function has [a default value](https://github.com/WordPress/gutenberg/blob/51e7afaaae029e625d7182660844216a8183d835/lib/compat/wordpress-6.2/formatting.php#L41-L53) for the regular expression on which it bases its calculations, but for now there is no filter hook.

## Tests

PHPUnit tests are implemented for the two functions.

- For the new core function `wp_word_count`: I created the exact same test cases for [the core tests implemented in qunit](https://github.com/WordPress/wordpress-develop/blob/trunk/tests/qunit/wp-admin/js/word-count.js).
- For `render_callback` function: I created test cases to verify that this block displays the correct text on the front end, depending on the amount of content.

talldan commented on packages/block-library/src/post-time-to-read/edit.js:
```diff
@@ -0,0 +1,106 @@
+		if ( minutesToRead !== 0 ) {
+			return sprintf(
+				/* translators: %d is the number of minutes the post will take to read. */
+				_n(
+					'You can read this post in %d minute.',
+					'You can read this post in %d minutes.',
+					minutesToRead
+				),
```
Thanks for all your works so far on this @t-hamano, it feels very close to ready now.

One concern I have is still around getting this text right. If the block is released in WordPress with this text it becomes hard to change in the future it because of backwards compatibility requirements.

The prefix / postfix option was tried, but based on feedback that doesn't seem preferable right now. If there was a desire to add it again in the future it'd be hard to do with the existing 'You can read this post in' text.

It'd be good to get your thoughts on this. I'm wondering if the text should be reduced down to just '%d minute' / '%d minutes'. That way users can mix it up with an icon (through an icon block from a plugin, core doesn't have one 😄 ), or another paragraph block, or even both.

javierarce replied:
I agree with the last idea: it will make this block more flexible and extensible, probably increasing its use by many more people.

t-hamano replied:
Thank you for the review, @talldan, @javierarce !

>  I'm wondering if the text should be reduced down to just '%d minute' / '%d minutes'. 

I think this is a very good idea! If we include "less than 1 minute" as a variation, it may not be possible to standardize the text before and after in all languages. Furthermore, the question of whether to capitalize the first letter of "less than" or lowercase arises.

As I mentioned in [this comment](https://github.com/WordPress/gutenberg/pull/43403#discussion_r991365917), I am concerned about the separation of paragraphs, but the freedom of variation would be convenient for users.

mtias replied:
Agreed, we should bring this down to just displaying the time and no extra wording around it.

talldan replied:
> I am concerned about the separation of paragraphs

Agree with that, and it came to my mind too. It wouldn't really be semantically correct to output paragraphs in a row like this, when the user is essentially trying to concatenate text. Screenreaders may also not announce the text properly.

I think it should be considered a separate problem to solve. There are likely more use cases beyond this block. I don't know if the dynamic tokens system is considered a solution to this, while I read the initial proposal, it sounds like the outcome has changed quite a bit.

There are other solutions too, like a variation of a paragraph that outputs a span, but that kind of thing would require users knowing about html.

t-hamano committed:
Show time only

t-hamano committed:
Update PHP unit test

t-hamano commented on .github/dependabot.yml:
```diff
@@ -2,9 +2,9 @@
-    schedule:
-      interval: "daily"
-    open-pull-requests-limit: 10
+    # Check for updates to GitHub Actions.
+    - package-ecosystem: 'github-actions'
+      directory: '/'
+      schedule:
+          interval: 'daily'
+      open-pull-requests-limit: 10
```
I believe this change was not intended, but was auto-formatted by lint-staged or something.
But I think the indentation and quoting is correct after the change.

talldan replied:
It's done that to me too. It looks good. If you want to move the change to a separate PR, I'll be happy to merge it.

t-hamano replied:
I submitted #45167 👍

t-hamano commented:
Updated: This block simply outputs only the text of the two variations of time (_1 minute_ / _x minutes_).

![capture](https://user-images.githubusercontent.com/54422211/196825532-e5bf6232-8a3d-472f-a2fe-2facf0844ddb.png)

t-hamano commented on lib/compat/wordpress-6.2/formatting.php:
```diff
@@ -0,0 +1,115 @@
+	 * }
+	 * @return int The word or character count.
+	 */
+	function wp_word_count( $text, $type, $settings = array() ) {
```
This function will be implemented in the core, but is this the correct directory to place the function?

talldan replied:
If you place it here it'll be backported when WordPress 6.2 comes around. The function should have the `gutenberg_` prefix while it lives in the plugin though and only be changed to `wordpress_` when it's moved to core. The backports happen manually and the release lead for 6.2 will check this folder. Here's the issue where backports for 6.1 were coordinated as an example - https://github.com/WordPress/gutenberg/issues/43440.

Alternatively you can open up a trac ticket and try to get it merged to core separately, but there's no real advantage to doing that. The only advantage would be if a core feature needs to use this function between now and 6.2.

t-hamano replied:
According to[ this comment](https://github.com/WordPress/gutenberg/pull/43403#discussion_r993617934), it appears that functions merged into core should have a wp_prefix, correct?

talldan replied:
I don't think that's right (cc @anton-vlasenko). If you check in the `lib/compat/wordpress-6.1` folder for example there are no `wp` prefixed functions, only `gutenberg` prefixed functions.

This code will be in the Gutenberg plugin for a few months before being merged into core around the time of the 6.2 beta. During the time it's in the plugin it needs a `gutenberg` prefix as the `wp` prefix is reserved for code in the WordPress core codebase.

When the 6.2 beta comes around, the function will need to be implemented in the WordPress core codebase. The WP core implementation will then have the `wp` prefix.

t-hamano replied:
In the `lib/compat` directory, there are several functions with the wp prefix.
The wp_word_count function in this case might be used for the same purpose?

- `lib/compat/wordpress-6.0/functions.php`: [wp_recursive_ksort](https://github.com/WordPress/gutenberg/blob/3c9a0207f51d731ff4d8cfbfa69daa16f797639f/lib/compat/wordpress-6.0/functions.php#L19)
- `lib/compat/wordpress-6.1/theme.php`: [wp_theme_get_element_class_name](https://github.com/WordPress/gutenberg/blob/3c9a0207f51d731ff4d8cfbfa69daa16f797639f/lib/compat/wordpress-6.1/theme.php#L44)
- `lib/compat/wordpress-6.2/get-global-styles-and-settings.php`: [wp_theme_has_theme_json](https://github.com/WordPress/gutenberg/blob/3c9a0207f51d731ff4d8cfbfa69daa16f797639f/lib/compat/wordpress-6.2/get-global-styles-and-settings.php#L16), [wp_theme_clean_theme_json_cached_data](https://github.com/WordPress/gutenberg/blob/3c9a0207f51d731ff4d8cfbfa69daa16f797639f/lib/compat/wordpress-6.2/get-global-styles-and-settings.php#L43)

talldan replied:
That's confusing to me. I always thought `wp` was reserved for core (Gutenberg code isn't part of core until a WordPress release and the code is backported).

I'll ask around.

ntsekouras replied:
In most cases I think @talldan is right. When we introduce a new function we need to `gutenberg` prefix it. When it lands in core we rename it(in core) to `wp` prefix. 

In this case I think it's okay, because it's used inside a `block/index.php` file and if I'm not wrong the functions there are stripped of `gutenberg` prefix for core.

talldan replied:
> In this case I think it's okay, because it's used inside a block/index.php file and if I'm not wrong the functions there are stripped of gutenberg prefix for core.

Ah yes, that might be why. 💡

t-hamano commented:
I have tested how the screen reader reads this block when combined with the paragraph block.

<details>
<summary>Block composition</summary>

Use the row block and add the paragraph block as prefix

![block-composition](https://user-images.githubusercontent.com/54422211/198813560-77ece35a-676c-455c-93cf-f583e1cdbdd6.png)

</details>

<details>
<summary>Screencast with audio</summary>

Use Narrator, Windows 11's default screen reader

https://user-images.githubusercontent.com/54422211/198813614-24b89446-b5c6-40bf-9cf3-c0385bff6ec8.mp4

</details>

t-hamano commented:
@mtias @talldan 
Is there anything I can do to help move this PR forward?

The biggest barrier to this block (as commented [here](https://github.com/WordPress/gutenberg/pull/43403#discussion_r1000084323)) is that when we want to combine it with a paragraph block to make a single sentence, the paragraph elements are split. However, so far there does not seem to be an ideal approach to this.

Do you have any good ideas whether we should release this block, treat it as experimental, or freeze it?

talldan commented:
My feeling is that it's ready to ship, I just haven't had time to do another code review recently. I'll try to prioritise it this week.

With the way it works at the moment, combining it with another sentence can be something that's iteratively improved through follow up PRs, though I'm not really sure on what the desired solution is for that.

masteradhoc commented:
@talldan would love to see this added soon :) thanks for the work everyone

talldan reviewed (approved):
This still works well, I rebased it locally, and there were no issues. Thanks for your hard work and patience. 

It's really nice that this works in a query loop as well as an individual post:
![Screen Shot 2023-01-31 at 4 15 04 pm](https://user-images.githubusercontent.com/677833/215704023-8a8c57bc-ac07-464d-91da-b9eec26edf89.png)

If it's merged prior to the next Gutenberg release (tomorrow) that means it will land in 6.2, and I don't see any reason that it shouldn't land. There are some backports to do, but the function should be fairly easy to port over. cc @ntsekouras @Mamaduka.

ntsekouras commented:
@talldan can you rebase and merge?

Mamaduka reviewed (commented):
Fantastic work, @t-hamano! The block looks good to land in the plugin.

Do we have an official policy for shipping new blocks with WP core? It might be a good idea to have this block tested in production via Gutenberg plugin for a few releases.

Time to read and word count calculations get tricky when working with non-Latin characters. E.g., Here's a long-standing bug in the Wordcount package #13796.

talldan commented:
I'm not sure if there's an official policy, but it might be an option to include it for the beta and then remove it later if we feel it's not ready.

Doing the reverse (not including it in the beta and then deciding we want it later) is a bit harder to wrangle.

ntsekouras commented:
>Time to read and word count calculations get tricky when working with non-Latin characters. E.g., Here's a long-standing bug in the Wordcount package https://github.com/WordPress/gutenberg/issues/13796.

I wasn't aware of this... Don't we already use the same though for the `outline` section in the editor?

talldan commented:
> I wasn't aware of this... Don't we already use the same though for the outline section in the editor?

Yep, I believe it's the same utility. Though this perhaps adds more exposure to the issue given the time to read will be shown on the front end of user's sites.

Maybe a way to mitigate it is to introduce a block setting that can change the way Time to Read is calculated. Though I'm not sure how well that would work with various i18n plugins and mixed language content.

It'd be nice if it worked automatically, that would probably entail progressing this issue - [Language cannot be set on posts](https://github.com/WordPress/gutenberg/issues/15350#top).

t-hamano commented:
I would be very happy if this block is included in the next WordPress 6.2, but I think it needs some more testing, especially in non-Latin languages.

How about releasing this PR as part of Gutenberg 15.2 and then gather more feedback with the goal of releasing it in the next WordPress 6.3?

If the results of that feedback indicate that it is not yet sufficient to merge into the core, I think we could mark it as experimental in `block.json`, as the table of contents block is. Alternatively, there may be a way to mark it as experimental at this time and remove the experimental mark when it is ready to be included in the WordPress core.

talldan commented:
My opinion is that it would still be ok to merge for beta and postpone the decision until there's more time to discuss it, but I'll leave that up to the release leads.

Using the experiment flag would omit the block from 6.2, but equally not merging until after the next Gutenberg RC would do the same.

> It'd be nice if it worked automatically, that would probably entail progressing this issue - https://github.com/WordPress/gutenberg/issues/15350#top.

I wonder if using site language to determine a strategy would be ok (though I guess language is not always the same as locale). Alternatively it might be possible to derive a strategy based on the unicode character ranges present in the written post content. There are a lot though - https://jrgraphix.net/r/Unicode/.

ntsekouras commented:
After the discussions here, I feel like it would be better to punt for 6.3.

>Using the experiment flag would omit the block from 6.2, but equally not merging until after the next Gutenberg RC would do the same.

Probably it's better to use the experimental flag and land. Thanks for all your work @t-hamano !

t-hamano committed:
Mark as experimental

t-hamano committed:
Move `wp_word_count()` function to `experimental` directory

t-hamano commented:
- ✅ Marked this block as experimental in 0dbee92.
- ✅ With this change, I have moved the `wp_word_count()` function from ` lib/compat/wordpress-6.2/formatting.php` to `lib/experimental/formatting.php` in 68c7e38.

If there are no problems with this change, I would like to merge.

Mamaduka commented:
A new `WP_Locale` property that might be useful for this block - https://github.com/WordPress/wordpress-develop/pull/4019.

t-hamano committed:
Use new function `wp_get_word_count_type()`

t-hamano commented:
Thanks for the info, @Mamaduka!

I discovered that a new `wp_get_word_count_type()` function was also added, so I replaced it with this function within the PHP rendering by 74cbb73.

At the same time, I added this function to `lib/experimental/l10n.php` for backward compatibility.

t-hamano commented:
Somehow, I am getting tons of errors with the Javascript Lint check...🤔

talldan commented:
> Somehow, I am getting tons of errors with the Javascript Lint check...🤔

It might be worth rebasing it against `trunk`.

t-hamano committed:
fix lint error

t-hamano commented:
I have fixed the lint error. It was simply an omission to update `tsconfig.json` 😅 It appears that the performance test failure is not due to this PR, but is being investigated in #48138.

t-hamano committed:
Merge branch 'trunk' into add/block-post-time-to-read

talldan commented:
Do you think this is ready to merge now @t-hamano?

t-hamano committed:
Merge branch 'trunk' into add/block-post-time-to-read

t-hamano committed:
Change namespace from `formatting` to `l10n`

t-hamano commented:
@talldan 
I think this PR is ready to merge.

✅ Merged the latest trunk and it is working as expected
✅ It works as expected with WordPress 6.0, which is supported by the Gutenberg plugin

However, I found that [wp_get_word_count_type()](https://github.com/WordPress/wordpress-develop/blob/12fcc5cfac7ba94278df45f88cdd7f4ef3e02864/src/wp-includes/l10n.php#L1829-L1838), introduced in WordPress 6.2, belongs to the `l10n` category. Therefore, in [ebc1d03](https://github.com/WordPress/gutenberg/pull/43403/commits/ebc1d03eec0e2929ef703f7731ab875182f9cb16), I have taken the following actions:

- Merge `lib/experimental/formatting.php` into `lib/experimental/l10n.php`
- Rename `phpunit/formatting-test.php` to `phpunit/l10n-test.php`

talldan commented:
Thanks for all your work on this, tests are green so I'll merge it 👍

bph commented:
I see that it was merged into the milestone 5.3 but I can't find it in the 5.3 RC1 changelog. Does this need to be backported to the Gutenberg RC or am I overlooking something? @richtabor @talldan ?

t-hamano commented:
@bph 
Perhaps it is hard to find because it is classified in the "Various" section.

![changelog](https://user-images.githubusercontent.com/54422211/223450398-4371e16f-abf5-4ad2-820e-a5dd8bf72b75.png)

bph commented:
Thank you so much! @t-hamano

bph commented:
@t-hamano thank you for all the hard work going through the various iterations. It's a great block. 
I think it works well without the prefix and suffix complications. 
I used it together with a paragraph block in my post template block and via the row block is displays fine after I added a block class and a custom CSS.  
Once the design tools, specifically the typography controls are added it, is quite flexible. 
![Screenshot 2023-03-08 at 16 30 14](https://user-images.githubusercontent.com/39980/223854758-2d9d23fa-0f0e-4056-a888-b6e761d2fe99.png)

t-hamano commented:
I have opened a ticket to add the `wp_word_count()` function implemented in this PR to the core:
https://core.trac.wordpress.org/ticket/57987#ticket
